File: src/lib/rg/hazards.ts
// Hazard domain service: serialization, geospatial context stats, priority recompute and
// DBSCAN cluster recomputation. Production parity: backend/app/services/{hazards,clustering,priority}.py
import { db } from "@/lib/db";
import type { Prisma } from "@prisma/client";
import { computePriority } from "./priority";
import { haversineM, inferRoadCriticality, nearestWard } from "./geo";
import { getSettings, type Settings } from "./settings";
import { dbscan, summarizeClusters } from "./dbscan";
import { ACTIONABLE_STATUSES, CLASS_META, DEFAULT_MAP_STATUSES, HAZARD_STATUSES, severityBandOf } from "./constants";
import { recommendedActionFor, riskFlagsFor } from "./risk";
import type { BboxDTO, HazardDTO, HazardClass, PriorityBand, ReportStatus, RiskFlagDTO, WorkOrderStatus } from "./types";

export const reportInclude = {
  media: true,
  detections: { orderBy: { confidence: "desc" as const } },
  priorityScores: { orderBy: { computedAt: "desc" as const }, take: 1 },
  workOrder: {
    select: {
      id: true,
      code: true,
      status: true,
      beforeMediaId: true,
      afterMediaId: true,
      resolutionNotes: true,
      completedAt: true,
      verifiedBy: true,
      verifiedAt: true,
      rejectReason: true,
    },
  },
  mergedReports: {
    select: {
      id: true,
      referenceCode: true,
      submitterName: true,
      createdAt: true,
      media: { select: { id: true, kind: true, mimeType: true } },
    },
  },
} satisfies Prisma.HazardReportInclude;

export type ReportWithRelations = Prisma.HazardReportGetPayload<{ include: typeof reportInclude }>;

export function serializeHazard(r: ReportWithRelations): HazardDTO {
  const ps = r.priorityScores[0];
  let priority: HazardDTO["priority"] = null;
  if (ps) {
    try {
      let riskFlags: RiskFlagDTO[] = [];
      if (ps.riskFlagsJson) {
        try { riskFlags = JSON.parse(ps.riskFlagsJson); } catch { riskFlags = []; }
      }
      priority = {
        score: ps.score,
        band: ps.band as PriorityBand,
        factors: JSON.parse(ps.explanationJson),
        riskFlags,
        recommendedAction: ps.recommendedAction ?? null,
        weights: JSON.parse(ps.weightsJson),
        overridden: ps.overridden,
        manualScore: ps.manualScore,
        computedAt: ps.computedAt.toISOString(),
      };
    } catch {
      priority = null;
    }
  }
  return {
    severityBand: severityBandOf(r.severity),
    reportCount: r.reportCount,
    uniqueReporters: r.uniqueReporters,
    lastReportedAt: r.lastReportedAt?.toISOString() ?? null,
    updatedAt: r.updatedAt.toISOString(),
    mergedEvidence: (r.mergedReports ?? []).flatMap((m) =>
      m.media.map((mm) => ({ id: mm.id, kind: mm.kind, mimeType: mm.mimeType, referenceCode: m.referenceCode, submitterName: m.submitterName, createdAt: m.createdAt.toISOString() }))
    ),
    id: r.id,
    referenceCode: r.referenceCode,
    hazardClass: r.hazardClass as HazardClass,
    severity: r.severity,
    status: r.status as ReportStatus,
    lat: r.lat,
    lng: r.lng,
    address: r.address,
    ward: r.ward,
    roadName: r.roadName,
    roadClass: (r.roadClass as HazardDTO["roadClass"]) ?? null,
    notes: r.notes,
    source: r.source,
    duplicateOfId: r.duplicateOfId,
    submitterName: r.submitterName,
    createdAt: r.createdAt.toISOString(),
    reviewedAt: r.reviewedAt?.toISOString() ?? null,
    reviewNote: r.reviewNote,
    userId: r.userId,
    media: r.media.map((m) => ({
      id: m.id,
      kind: m.kind,
      mimeType: m.mimeType,
      width: m.width,
      height: m.height,
      durationSec: m.durationSec,
    })),
    detections: r.detections.map(
      (d): BboxDTO => ({
        id: d.id,
        hazardClass: d.hazardClass as HazardClass,
        confidence: d.confidence,
        bbox: [d.bboxX, d.bboxY, d.bboxW, d.bboxH],
        areaRatio: d.areaRatio,
        severity: d.severity,
        engine: d.engine,
        modelVersion: d.modelVersion,
      })
    ),
    priority,
    clusterId: r.clusterId,
    workOrderStatus: (r.workOrder?.status as WorkOrderStatus) ?? null,
    workOrder: r.workOrder
      ? {
          id: r.workOrder.id,
          code: r.workOrder.code,
          status: r.workOrder.status as WorkOrderStatus,
          beforeMediaId: r.workOrder.beforeMediaId,
          afterMediaId: r.workOrder.afterMediaId,
          resolutionNotes: r.workOrder.resolutionNotes,
          completedAt: r.workOrder.completedAt?.toISOString() ?? null,
          verifiedBy: r.workOrder.verifiedBy,
          verifiedAt: r.workOrder.verifiedAt?.toISOString() ?? null,
          rejectReason: r.workOrder.rejectReason,
        }
      : null,
  };
}

/** Same-class actionable reports within 120 m over 60 days (density) and 75 m over 30 days (recurrence). */
export async function computeContextStats(params: {
  lat: number;
  lng: number;
  hazardClass: string;
  excludeReportId?: string;
}): Promise<{ neighborCount: number; recurrenceCount: number }> {
  const actionable: Prisma.HazardReportWhereInput = {
    status: { in: ACTIONABLE_STATUSES },
    duplicateOfId: null,
  };
  const around = {
    ...actionable,
    lat: { gte: params.lat - 0.003, lte: params.lat + 0.003 },
    lng: { gte: params.lng - 0.004, lte: params.lng + 0.004 },
    ...(params.excludeReportId ? { id: { not: params.excludeReportId } } : {}),
  };
  const near60 = await db.hazardReport.findMany({
    where: { ...around, hazardClass: params.hazardClass, createdAt: { gte: new Date(Date.now() - 60 * 86400_000) } },
    select: { id: true, lat: true, lng: true },
  });
  const near30 = await db.hazardReport.findMany({
    where: { ...around, hazardClass: params.hazardClass, createdAt: { gte: new Date(Date.now() - 30 * 86400_000) } },
    select: { id: true, lat: true, lng: true },
  });
  const p = { lat: params.lat, lng: params.lng };
  const neighborCount = near60.filter((r) => haversineM(p, r) <= 120).length;
  const recurrenceCount = near30.filter((r) => haversineM(p, r) <= 75).length;
  return { neighborCount, recurrenceCount };
}

export async function recomputePriorityForReport(reportId: string, settings?: Settings): Promise<void> {
  const s = settings ?? (await getSettings());
  const report = await db.hazardReport.findUnique({
    where: { id: reportId },
    include: {
      workOrder: { select: { status: true } },
      priorityScores: { orderBy: { computedAt: "desc" }, take: 1 },
      media: { where: { kind: "ORIGINAL" }, take: 1 },
      detections: { orderBy: { confidence: "desc" }, take: 1 },
    },
  });
  if (!report) return;
  if (report.status === "REJECTED" || report.status === "MERGED") return;

  const { neighborCount, recurrenceCount } = await computeContextStats({
    lat: report.lat,
    lng: report.lng,
    hazardClass: report.hazardClass,
    excludeReportId: report.id,
  });
  const criticality = report.roadCriticality ?? inferRoadCriticality(report.roadName, report.roadClass);
  const ageDays = (Date.now() - report.createdAt.getTime()) / 86400_000;
  const prev = report.priorityScores[0];
  const resolved = report.workOrder?.status === "COMPLETED" || report.workOrder?.status === "VERIFICATION_PENDING" || report.workOrder?.status === "VERIFIED" || report.workOrder?.status === "CLOSED";
  const result = computePriority({
    severity: report.severity,
    neighborCount,
    roadCriticality: criticality,
    recurrenceCount,
    ageDays,
    resolved,
    weights: s.weights,
    overridden: prev?.overridden ?? false,
    manualScore: prev?.manualScore ?? null,
  });

  // explainable risk flags from real signals (AI confidence, bbox extent, community reports, road class)
  const top = report.detections[0] ?? null;
  const riskFlags = riskFlagsFor({
    hazardClass: report.hazardClass,
    confidence: top?.confidence ?? null,
    areaRatio: top?.areaRatio ?? null,
    severity: report.severity,
    roadClass: report.roadClass,
    roadName: report.roadName,
    recurrenceCount,
    neighborCount,
    reportCount: report.reportCount,
    uniqueReporters: report.uniqueReporters,
  });
  const recommendedAction = recommendedActionFor(result.band, s.actions);

  const shared = {
    score: result.score,
    band: result.band,
    severityNorm: result.factors[0].normalized,
    densityNorm: result.factors[1].normalized,
    criticalityNorm: result.factors[2].normalized,
    recurrenceNorm: result.factors[3].normalized,
    ageNorm: result.factors[4].normalized,
    weightsJson: JSON.stringify(result.weights),
    explanationJson: JSON.stringify(result.factors),
    riskFlagsJson: JSON.stringify(riskFlags),
    recommendedAction,
    computedAt: new Date(),
  };
  await db.priorityScore.upsert({
    where: { reportId },
    update: shared,
    create: {
      reportId,
      ...shared,
      overridden: prev?.overridden ?? false,
      overriddenBy: prev?.overriddenBy ?? null,
      manualScore: prev?.manualScore ?? null,
    },
  });
}

export interface RecomputeScope {
  bbox?: { minLat: number; minLng: number; maxLat: number; maxLng: number };
  sinceDays?: number;
  epsM?: number;
  minPts?: number;
}

export interface RecomputeSummary {
  scopedReports: number;
  clusters: number;
  assignedReports: number;
  epsM: number;
  minPts: number;
  computedAt: string;
}

let recomputeInFlight = false;

/** Full clustering recompute (idempotent): DBSCAN → clusters + memberships → priority refresh. */
export async function recomputeClusters(scope: RecomputeScope = {}): Promise<RecomputeSummary> {
  if (recomputeInFlight) throw new Error("Cluster recompute already in progress");
  recomputeInFlight = true;
  try {
    const s = await getSettings();
    const epsM = scope.epsM ?? s.cluster.epsM;
    const minPts = scope.minPts ?? s.cluster.minPts;
    const sinceDays = scope.sinceDays ?? s.cluster.sinceDays;

    const where: Prisma.HazardReportWhereInput = {
      status: { in: ACTIONABLE_STATUSES },
      duplicateOfId: null,
      createdAt: { gte: new Date(Date.now() - sinceDays * 86400_000) },
      ...(scope.bbox
        ? { lat: { gte: scope.bbox.minLat, lte: scope.bbox.maxLat }, lng: { gte: scope.bbox.minLng, lte: scope.bbox.maxLng } }
        : {}),
    };
    const reports = await db.hazardReport.findMany({ where, select: { id: true, lat: true, lng: true } });
    const labels = dbscan(reports.map((r) => ({ lat: r.lat, lng: r.lng })), epsM, minPts);
    const summaries = summarizeClusters(reports.map((r) => ({ lat: r.lat, lng: r.lng })), labels);

    // gather aggregate info per cluster
    const full = await db.hazardReport.findMany({
      where: { id: { in: reports.map((r) => r.id) } },
      select: { id: true, hazardClass: true, severity: true, ward: true, lat: true, lng: true },
    });
    const byId = new Map(full.map((r) => [r.id, r]));

    await db.$transaction([
      db.clusterMembership.deleteMany({}),
      db.hazardCluster.deleteMany({}),
    ]);

    let assigned = 0;
    for (let ci = 0; ci < summaries.length; ci++) {
      const sum = summaries[ci];
      const members = sum.memberIndexes.map((i) => byId.get(reports[i].id)!).filter(Boolean);
      const classTally = new Map<string, number>();
      members.forEach((m) => classTally.set(m.hazardClass, (classTally.get(m.hazardClass) ?? 0) + 1));
      const dominant = [...classTally.entries()].sort((a, b) => b[1] - a[1])[0]?.[0] ?? "pothole";
      const avgSeverity = members.reduce((acc, m) => acc + m.severity, 0) / Math.max(1, members.length);
      const maxSeverity = members.reduce((acc, m) => Math.max(acc, m.severity), 0);
      const ward = nearestWard({ lat: sum.centerLat, lng: sum.centerLng });
      const cluster = await db.hazardCluster.create({
        data: {
          label: ward ? `C-${ci + 1} · ${ward}` : `C-${ci + 1}`,
          centerLat: sum.centerLat,
          centerLng: sum.centerLng,
          radiusM: Math.round(sum.radiusM),
          hazardCount: members.length,
          dominantClass: dominant,
          avgSeverity: Math.round(avgSeverity * 100) / 100,
          maxSeverity,
          paramsJson: JSON.stringify({ epsM, minPts, sinceDays }),
        },
      });
      for (const m of members) {
        const point = { lat: m.lat, lng: m.lng };
        const dist = haversineM(point, { lat: sum.centerLat, lng: sum.centerLng });
        await db.clusterMembership.create({
          data: { clusterId: cluster.id, reportId: m.id, distanceM: Math.round(dist) },
        });
        await db.hazardReport.update({ where: { id: m.id }, data: { clusterId: cluster.id } });
        assigned++;
      }
    }
    // clear stale cluster pointers for in-scope noise points
    const memberIds = new Set(summaries.flatMap((s) => s.memberIndexes.map((i) => reports[i].id)));
    for (const r of reports) {
      if (!memberIds.has(r.id) && r.id) {
        await db.hazardReport.updateMany({ where: { id: r.id, clusterId: { not: null } }, data: { clusterId: null } });
      }
    }

    // refresh priorities for the whole scope (density may have changed)
    for (const r of reports) await recomputePriorityForReport(r.id, s);

    return {
      scopedReports: reports.length,
      clusters: summaries.length,
      assignedReports: assigned,
      epsM,
      minPts,
      computedAt: new Date().toISOString(),
    };
  } finally {
    recomputeInFlight = false;
  }
}

/** Lazily refresh clusters before serving map data (bounded to once per 10 minutes). */
let lastAutoRecompute = 0;
export async function autoRecomputeIfStale(): Promise<void> {
  const s = await getSettings();
  if (!s.autoRecompute) return;
  if (Date.now() - lastAutoRecompute < 10 * 60_000) return;
  lastAutoRecompute = Date.now();
  try {
    await recomputeClusters();
  } catch (err) {
    console.error("[clusters] auto-recompute failed", err);
  }
}

/** Parses shared filter query params for /api/hazards and /api/map. */
export function parseHazardFilters(sp: URLSearchParams): Prisma.HazardReportWhereInput {
  const where: Prisma.HazardReportWhereInput = { duplicateOfId: null };
  const classes = sp.get("classes");
  if (classes) where.hazardClass = { in: classes.split(",").filter((c) => validClass(c)) };
  const sevMin = Number(sp.get("severityMin") ?? 1);
  const sevMax = Number(sp.get("severityMax") ?? 5);
  if (sevMin > 1 || sevMax < 5) where.severity = { gte: Math.max(1, sevMin), lte: Math.min(5, sevMax) };
  const statuses = sp.get("statuses");
  where.status = statuses
    ? { in: statuses.split(",").filter((s) => (VALID_STATUSES as string[]).includes(s)) as ReportStatus[] }
    : { in: DEFAULT_MAP_STATUSES };
  const from = sp.get("from");
  const to = sp.get("to");
  if (from || to) where.createdAt = { ...(from ? { gte: new Date(from) } : {}), ...(to ? { lte: new Date(`${to.slice(0, 10)}T23:59:59.999Z`) } : {}) };
  const bbox = sp.get("bbox");
  if (bbox) {
    const [minLng, minLat, maxLng, maxLat] = bbox.split(",").map(Number);
    if ([minLat, minLng, maxLat, maxLng].every(Number.isFinite)) {
      where.lat = { gte: Math.min(minLat, maxLat), lte: Math.max(minLat, maxLat) };
      where.lng = { gte: Math.min(minLng, maxLng), lte: Math.max(minLng, maxLng) };
    }
  }
  const q = sp.get("q");
  if (q) where.OR = [{ referenceCode: { contains: q } }, { address: { contains: q } }, { ward: { contains: q } }, { roadName: { contains: q } }];
  const bands = sp.get("bands");
  if (bands) {
    const bandList = bands.split(",").filter((b) => VALID_BANDS.includes(b));
    if (bandList.length > 0) where.priorityScores = { some: { band: { in: bandList } } };
  }
  return where;
}

const VALID_STATUSES = HAZARD_STATUSES;
const VALID_BANDS = ["CRITICAL", "HIGH", "MODERATE", "LOW"];

function validClass(c: string): c is HazardClass {
  return (Object.keys(CLASS_META) as string[]).includes(c);
}

File: backend/app/services/hazards.py
"""Hazard domain service — parity with src/lib/rg/hazards.ts.

Covers: serialization (HazardDTO shape), geospatial context stats (density /
recurrence), priority recompute (upsert with override preservation) and the
DBSCAN cluster recompute (delete + recreate in scope, idempotent).
"""
from __future__ import annotations

import json
import threading
from datetime import timedelta
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.models import (
    ClusterMembership,
    HazardCluster,
    HazardReport,
    PriorityScore,
    WorkOrder,
    utcnow,
)
from app.services.geo import (
    haversine_m,
    infer_road_criticality,
    nearest_ward,
)
from app.services.priority import PriorityInput, compute_priority
from app.services.severity import HAZARD_CLASSES as HAZARD_CLASSES_ALL
from app.services.severity import area_ratio_of, severity_from_detection
from app.services.settings import get_settings

ACTIONABLE_STATUSES = ("PENDING_REVIEW", "APPROVED", "FLAGGED")
VALID_STATUSES = ("PENDING_REVIEW", "APPROVED", "REJECTED", "MERGED", "FLAGGED")
VALID_BANDS = ("CRITICAL", "HIGH", "MEDIUM", "LOW")

REFERENCE_ALPHABET = "ABCDEFGHJKMNPQRSTUVWXYZ23456789"


def make_reference() -> str:
    """`RG-XXXXXX` — ambiguity-free alphabet (no I/L/O/0/1)."""
    import secrets

    code = "".join(secrets.choice(REFERENCE_ALPHABET) for _ in range(6))
    return f"RG-{code}"


# ------------------------------------------------------------------ serialization
def serialize_hazard(r: HazardReport, work_order_status: str | None = None) -> dict[str, Any]:
    priority: dict[str, Any] | None = None
    ps = r.priority_score
    if ps is not None:
        try:
            priority = {
                "score": ps.score,
                "band": ps.band,
                "factors": json.loads(ps.explanation_json),
                "weights": json.loads(ps.weights_json),
                "overridden": ps.overridden,
                "manualScore": ps.manual_score,
                "computedAt": ps.computed_at.isoformat(),
            }
        except (TypeError, ValueError):
            priority = None

    return {
        "id": r.id,
        "referenceCode": r.reference_code,
        "hazardClass": r.hazard_class,
        "severity": r.severity,
        "status": r.status,
        "lat": r.lat,
        "lng": r.lng,
        "address": r.address,
        "ward": r.ward,
        "roadName": r.road_name,
        "roadClass": r.road_class,
        "notes": r.notes,
        "source": r.source,
        "duplicateOfId": r.duplicate_of_id,
        "createdAt": r.created_at.isoformat(),
        "reviewedAt": r.reviewed_at.isoformat() if r.reviewed_at else None,
        "reviewNote": r.review_note,
        "userId": r.user_id,
        "media": [
            {
                "id": m.id,
                "kind": m.kind,
                "mimeType": m.mime_type,
                "width": m.width,
                "height": m.height,
                "durationSec": m.duration_sec,
            }
            for m in r.media
        ],
        "detections": [
            {
                "id": d.id,
                "hazardClass": d.hazard_class,
                "confidence": d.confidence,
                "bbox": (d.bbox_x, d.bbox_y, d.bbox_w, d.bbox_h),
                "areaRatio": d.area_ratio,
                "severity": d.severity,
                "engine": d.engine,
                "modelVersion": d.model_version,
            }
            for d in r.detections
        ],
        "priority": priority,
        "clusterId": r.cluster_id,
        "workOrderStatus": work_order_status,
    }


def load_report(db: Session, report_id: str) -> HazardReport | None:
    stmt = (
        select(HazardReport)
        .where(HazardReport.id == report_id)
        .options(
            selectinload(HazardReport.media),
            selectinload(HazardReport.detections),
            selectinload(HazardReport.priority_score),
        )
    )
    return db.execute(stmt).scalar_one_or_none()


def work_order_status_for(db: Session, report_id: str) -> str | None:
    row = db.execute(select(WorkOrder.status).where(WorkOrder.hazard_report_id == report_id)).scalar_one_or_none()
    return row


# ----------------------------------------------------------- context statistics
def compute_context_stats(
    db: Session,
    *,
    lat: float,
    lng: float,
    hazard_class: str,
    exclude_report_id: str | None = None,
) -> dict[str, int]:
    """Same-class actionable reports within 120 m / 60 d (density) and 75 m / 30 d (recurrence)."""
    from app.services.geo import neighbors_within

    base = [
        HazardReport.status.in_(ACTIONABLE_STATUSES),
        HazardReport.duplicate_of_id.is_(None),
        HazardReport.hazard_class == hazard_class,
    ]
    if exclude_report_id:
        base.append(HazardReport.id != exclude_report_id)

    near60 = neighbors_within(
        db, HazardReport, lat, lng, 120, (*base, HazardReport.created_at >= utcnow() - timedelta(days=60))
    )
    near30 = neighbors_within(
        db, HazardReport, lat, lng, 75, (*base, HazardReport.created_at >= utcnow() - timedelta(days=30))
    )
    return {"neighborCount": len(near60), "recurrenceCount": len(near30)}


# ------------------------------------------------------------ priority recompute
def recompute_priority_for_report(db: Session, report_id: str, settings_bundle: dict | None = None) -> None:
    bundle = settings_bundle or get_settings(db)
    report = load_report(db, report_id)
    if report is None:
        return
    if report.status in ("REJECTED", "MERGED"):
        return

    stats = compute_context_stats(
        db, lat=report.lat, lng=report.lng, hazard_class=report.hazard_class, exclude_report_id=report.id
    )
    criticality = report.road_criticality
    if criticality is None:
        criticality = infer_road_criticality(report.road_name, report.road_class)
    age_days = max(0.0, (utcnow() - report.created_at.replace(tzinfo=None)).total_seconds() / 86400)

    wo_status = work_order_status_for(db, report.id)
    prev = report.priority_score

    result = compute_priority(
        PriorityInput(
            severity=report.severity,
            neighbor_count=stats["neighborCount"],
            road_criticality=float(criticality),
            recurrence_count=stats["recurrenceCount"],
            age_days=age_days,
            resolved=wo_status == "RESOLVED",
            weights=bundle["weights"],
            overridden=bool(prev and prev.overridden),
            manual_score=prev.manual_score if prev else None,
        ),
        now=utcnow(),
    )

    norms = result.norms
    if prev is None:
        db.add(
            PriorityScore(
                report_id=report.id,
                score=result.score,
                band=result.band,
                severity_norm=norms["severity"],
                density_norm=norms["density"],
                criticality_norm=norms["criticality"],
                recurrence_norm=norms["recurrence"],
                age_norm=norms["age"],
                weights_json=json.dumps(result.weights),
                explanation_json=json.dumps([f.as_dict() for f in result.factors]),
                overridden=result.overridden,
                overridden_by=None,
                manual_score=result.manual_score,
                computed_at=utcnow(),
            )
        )
    else:
        prev.score = result.score
        prev.band = result.band
        prev.severity_norm = norms["severity"]
        prev.density_norm = norms["density"]
        prev.criticality_norm = norms["criticality"]
        prev.recurrence_norm = norms["recurrence"]
        prev.age_norm = norms["age"]
        prev.weights_json = json.dumps(result.weights)
        prev.explanation_json = json.dumps([f.as_dict() for f in result.factors])
        prev.computed_at = utcnow()
    db.flush()


# --------------------------------------------------------------- cluster recompute
def cluster_label(index: int, ward: str) -> str:
    return f"C-{index + 1} · {ward}"


_recompute_lock = threading.Lock()


class RecomputeInProgress(RuntimeError):
    pass


def recompute_clusters(
    db: Session,
    *,
    bbox: dict[str, float] | None = None,
    since_days: int | None = None,
    eps_m: float | None = None,
    min_pts: int | None = None,
) -> dict[str, Any]:
    """Full clustering recompute — idempotent (delete + recreate in scope).

    Scope: actionable (PENDING_REVIEW/APPROVED/FLAGGED), non-duplicate reports
    within `since_days` and optionally inside a bbox. After re-clustering, every
    in-scope report's priority is refreshed (density may have changed).
    """
    from app.services.clustering import dbscan, summarize_clusters

    if not _recompute_lock.acquire(blocking=False):
        raise RecomputeInProgress("Cluster recompute already in progress")
    try:
        bundle = get_settings(db)
        eps = float(eps_m if eps_m is not None else bundle["cluster"]["epsM"])
        min_pts_v = int(min_pts if min_pts is not None else bundle["cluster"]["minPts"])
        window_days = int(since_days if since_days is not None else bundle["cluster"]["sinceDays"])

        stmt = select(HazardReport).where(
            HazardReport.status.in_(ACTIONABLE_STATUSES),
            HazardReport.duplicate_of_id.is_(None),
            HazardReport.created_at >= utcnow() - timedelta(days=window_days),
        )
        if bbox:
            stmt = stmt.where(
                HazardReport.lat >= bbox["minLat"],
                HazardReport.lat <= bbox["maxLat"],
                HazardReport.lng >= bbox["minLng"],
                HazardReport.lng <= bbox["maxLng"],
            )
        reports = list(db.execute(stmt).scalars().all())
        points = [(r.lat, r.lng) for r in reports]
        labels = dbscan(points, eps_m=eps, min_pts=min_pts_v)
        summaries = summarize_clusters(points, labels)

        # --- idempotent delete: memberships + clusters touching the scope
        scoped_ids = [r.id for r in reports]
        if scoped_ids:
            old_memberships = db.execute(
                select(ClusterMembership).where(ClusterMembership.report_id.in_(scoped_ids))
            ).scalars().all()
            touched_cluster_ids = {m.cluster_id for m in old_memberships}
            for m in old_memberships:
                db.delete(m)
            # clear stale pointers
            db.execute(
                HazardReport.__table__.update()
                .where(HazardReport.id.in_(scoped_ids), HazardReport.cluster_id.isnot(None))
                .values(cluster_id=None)
            )
            # drop clusters that no longer have any member
            for cluster_id in touched_cluster_ids:
                remaining = db.execute(
                    select(ClusterMembership.id).where(ClusterMembership.cluster_id == cluster_id).limit(1)
                ).scalar_one_or_none()
                if remaining is None:
                    cluster = db.get(HazardCluster, cluster_id)
                    if cluster is not None:
                        db.delete(cluster)
        db.flush()

        # --- recreate
        by_id = {r.id: r for r in reports}
        assigned = 0
        for ci, summary in enumerate(summaries):
            members = [by_id[scoped_ids[i]] for i in summary["memberIndexes"]]
            class_tally: dict[str, int] = {}
            for m in members:
                class_tally[m.hazard_class] = class_tally.get(m.hazard_class, 0) + 1
            dominant = max(class_tally.items(), key=lambda kv: kv[1])[0] if class_tally else "pothole"
            avg_severity = sum(m.severity for m in members) / max(1, len(members))
            max_severity = max(m.severity for m in members) if members else 0
            ward = nearest_ward(summary["centerLat"], summary["centerLng"])

            cluster = HazardCluster(
                label=cluster_label(ci, ward),
                center_lat=summary["centerLat"],
                center_lng=summary["centerLng"],
                center=f"{summary['centerLat']:.7f},{summary['centerLng']:.7f}",  # "lat,lng" → WKT via GeographyPoint
                radius_m=round(summary["radiusM"]),
                hazard_count=len(members),
                dominant_class=dominant,
                avg_severity=round(avg_severity * 100) / 100,
                max_severity=max_severity,
                params_json=json.dumps({"epsM": eps, "minPts": min_pts_v, "sinceDays": window_days}),
                computed_at=utcnow(),
            )
            db.add(cluster)
            db.flush()
            for m in members:
                dist = haversine_m(m.lat, m.lng, summary["centerLat"], summary["centerLng"])
                db.add(ClusterMembership(cluster_id=cluster.id, report_id=m.id, distance_m=round(dist)))
                m.cluster_id = cluster.id
                assigned += 1
        db.flush()

        # refresh priorities for the whole scope (density may have changed)
        for r in reports:
            recompute_priority_for_report(db, r.id, bundle)
        db.commit()

        return {
            "scopedReports": len(reports),
            "clusters": len(summaries),
            "assignedReports": assigned,
            "epsM": eps,
            "minPts": min_pts_v,
            "computedAt": utcnow().isoformat(),
        }
    finally:
        _recompute_lock.release()


# --------------------------------------------------------------------- auto mode
_last_auto_recompute = 0.0
AUTO_RECOMPUTE_MIN_INTERVAL = 10 * 60  # bounded to once per 10 minutes


def auto_recompute_if_stale(db: Session) -> None:
    """Lazily refresh clusters before serving map data (map.autoRecompute=true)."""
    global _last_auto_recompute
    import time

    bundle = get_settings(db)
    if not bundle["autoRecompute"]:
        return
    if time.monotonic() - _last_auto_recompute < AUTO_RECOMPUTE_MIN_INTERVAL:
        return
    _last_auto_recompute = time.monotonic()
    try:
        recompute_clusters(db)
    except RecomputeInProgress:
        pass
    except Exception:  # noqa: BLE001 — auto mode must never break map serving
        import logging

        logging.getLogger("roadguard.clusters").exception("auto-recompute failed")


# ------------------------------------------------------------------ review utils
def ai_severity_for_report(db: Session, report: HazardReport, hazard_class: str) -> int:
    """AI-derived severity refined by duplicate context (used when user severity absent)."""
    detections = sorted(report.detections, key=lambda d: d.confidence, reverse=True)
    if not detections:
        return report.severity
    primary = detections[0]
    area = primary.area_ratio or area_ratio_of((primary.bbox_x, primary.bbox_y, primary.bbox_w, primary.bbox_h))
    stats = compute_context_stats(db, lat=report.lat, lng=report.lng, hazard_class=hazard_class)
    return severity_from_detection(
        hazard_class=hazard_class,
        confidence=primary.confidence,
        area_ratio=area,
        duplicate_count=stats["recurrenceCount"],
    )


__all__ = [
    "ACTIONABLE_STATUSES",
    "VALID_BANDS",
    "VALID_STATUSES",
    "RecomputeInProgress",
    "ai_severity_for_report",
    "cluster_label",
    "compute_context_stats",
    "load_report",
    "make_reference",
    "recompute_clusters",
    "recompute_priority_for_report",
    "serialize_hazard",
    "work_order_status_for",
]
